### DISCUSSION TITLE
MPICH symbols cannot be dlsym'd after `import MPI`?

### DISCUSSION THREAD

**Discussion by leofang** [mpi4py/mpi4py#285] [General]
Reproducer:
```cython
# test_dlsym.pyx 
# compile with "cythonize -i -3 test_dlsym.pyx"
cdef extern from "<dlfcn.h>" nogil:
    void *dlsym(void *handle, const char *symbol)
    void* DEFAULT 'RTLD_DEFAULT'
    #void* NEXT 'RTLD_NEXT'

cpdef load_symbol():
    cdef char* name = "MPI_Comm_rank"
    cdef void* symbol = dlsym(DEFAULT, name)
    if symbol == NULL:
        raise RuntimeError("dlsym with default failed")
```
This works with Open MPI:
```python
>>> from mpi4py import MPI  # init
>>> import test_dlsym
>>> test_dlsym.load_symbol()  # MPI_Comm_rank can be loaded because libmpi.so is loaded with global scope
>>>
```
but not with MPICH:
```python
>>> from mpi4py import MPI  # init
>>> import test_dlsym
>>> test_dlsym.load_symbol()  # MPI_Comm_rank cannot be loaded because libmpi.so is loaded with local scope
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "test_dlsym.pyx", line 7, in test_dlsym.load_symbol
    cpdef load_symbol():
  File "test_dlsym.pyx", line 11, in test_dlsym.load_symbol
    raise RuntimeError("dlsym with default failed")
RuntimeError: dlsym with default failed
```

An obvious workaround is to "reload" `libmpi.so` with the global scope after `import MPI`:
```python
>>> import ctypes
>>> so = ctypes.CDLL("libmpi.so", ctypes.RTLD_GLOBAL)
>>> test_dlsym.load_symbol()
>>> 
```
but still this MPI-dependent behavior is a bit surprising. I haven't tested if this is a regression or not.

**Comment by leofang** [comment]
What's puzzling me is I don't understand why `libmpi.so` would be loaded with the local scope for the case of MPICH. I looked at the codebase and didn't find anything unusual. It seems the `MPI` module is linked to `libmpi.so` as usual, nothing suspicious.

**Comment by leofang** [comment]
> What's puzzling me is I don't understand why libmpi.so would be loaded with the local scope for the case of MPICH.

Looks like it's because of CPython (`ctypes.DEFAULT_MODE` is `ctypes.RTLD_LOCAL`)? This is another workaround for MPICH:
```python
>>> import os, sys
>>> sys.setdlopenflags(os.RTLD_LAZY | os.RTLD_GLOBAL)
>>> from mpi4py import MPI
>>> import test_dlsym
>>> test_dlsym.load_symbol()
>>>
```
and I guess Open MPI works differently due to the compatibility layer doing `dlopen`? https://github.com/mpi4py/mpi4py/blob/master/src/lib-mpi/compat/openmpi.h

**Comment by dalcinl** [comment]
May I ask why are you looking into these details? What other nasty issue are you trying to resolve/investigate?

> and I guess Open MPI works differently due to the compatibility layer doing `dlopen`?

Exactly, but note that these hacks are not active for `openmpi>=4.0.0`. Are you actually using openmpi v3.x or older?

**Comment by leofang** [comment]
> May I ask why are you looking into these details? What other nasty issue are you trying to resolve/investigate?

We have a plugin-like system where it assumes MPI symbols can be dlsym'd after MPI_Init is called. This assumption works for C/C++ executables linked with -lmpi, but not with Python + MPICH.

> Exactly, but note that these hacks are not active for `openmpi>=4.0.0`. Are you actually using openmpi v3.x or older?

This is surprising, I tested everything with the conda packages without specifying the version, so it certainly brought the latest openmpi (4.1.4 IIRC).

**Comment by dalcinl** [comment]
> We have a plugin-like system where it assumes MPI symbols can be dlsym'd after MPI_Init is called. This assumption works for C/C++ executables linked with -lmpi, 

Of course, the dynamic linker runs executables by loading dependent libraries with global symbol resolution.

> but not with Python + MPICH.

Indeed. Python loads ext modules with local symbol resolution by default. In mpi4py, for some MPI implementations like openmpi<4, I had no other option than adding the re-dlopen'ing of the shared library to make things work. But you know I'm not a big fan of breaking conventions and ignoring default behavior, so I only do it if extremely necessary.  And this was never necessary for MPICH.

> This is surprising, I tested everything with the conda packages without specifying the version, so it certainly brought the latest openmpi (4.1.4 IIRC).

So perhaps openmpi v4 (or one of its plugins) is actually re-`dlopen`'ing its own `libmpi.so` at `MPI_Init` time? Otherwise, I cannot explain this behavior. Can you try running your dlsym tests without invoking `MPI_Init`?

**Comment by leofang** [comment]
> Can you try running your dlsym tests without invoking `MPI_Init`?

@dalcinl like this?

```python
>>> import mpi4py
>>> mpi4py.rc.initialize = False
>>> from mpi4py import MPI
>>> import test_dlsym
>>> test_dlsym.load_symbol() 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "test_dlsym.pyx", line 7, in test_dlsym.load_symbol
    cpdef load_symbol():
  File "test_dlsym.pyx", line 11, in test_dlsym.load_symbol
    raise RuntimeError("dlsym with default failed")
RuntimeError: dlsym with default failed
```

> But you know I'm not a big fan of breaking conventions and ignoring default behavior, so I only do it if extremely necessary. And this was never necessary for MPICH.

Makes sense indeed.

**Comment by dalcinl** [comment]
> @dalcinl like this?

Sorry, my recommendation makes no sense. mpi4py's `dlopen` hackery actually works at `MPI_Init` time, so if you don't call it, then the library is not re-opened with `RTLD_GLOBAL` visibility, and your `dlsym` test should always fail.

Also, note that I made a mistake in all my previous comments, mpi4py with openmpi uses the `dlopen` hack for `openmpi<3`. Of course, this mistake is inconsequential, I'm still puzzled, I don't get why your test works with openmpi. 

Can you try running your test under the `strace` or `ltrace` commands? The idea is to figure out where `dlopen` calls to `libmpi.so` happen.

I any case, I still don't have a good suggestion about how to make your plugin work flawlessly. Tell me something... is your plugin Linux-only or you need macOS support as well?

**Comment by leofang** [comment]
Let me convert this to a discussion since this is neither a bug nor a feature request 🙂

**Comment by leofang** [comment]
> Of course, this mistake is inconsequential, I'm still puzzled, I don't get why your test works with openmpi.

Isn't it simply because Open MPI is re-loaded in the global scope? 

> Tell me something... is your plugin Linux-only or you need macOS support as well?

It's Linux-only. I don't think mpi4py should sweat and make any change. Given there are plenty of workarounds we can just suggest MPICH users (if any) to apply one of them.

**Reply by dalcinl** [reply]
> Isn't it simply because Open MPI is re-loaded in the global scope?

Yes, but who is doing that??? If you are using openmpi>=3, then it should not be mpi4py as the re-dlopening should not happen for that version range. Is Open MPI re-loading itself?

**Reply by leofang** [reply]
Wait, hold on, I thought you said mpi4py is the one doing it?
> mpi4py with openmpi uses the dlopen hack for openmpi>=3.

**Reply by dalcinl** [reply]
I messed up again, I should have said `openmpi<3` (I've edited my previous comment)

**Reply by dalcinl** [reply]
[Use the source, Luke!](https://github.com/mpi4py/mpi4py/blob/master/src/lib-mpi/compat/openmpi.h#L16)

**Reply by leofang** [reply]
I did `strace python test.py` but I don't see anything unusual, am I missing something?
[strace.txt](https://github.com/mpi4py/mpi4py/files/10179774/strace.txt)

**Reply by dalcinl** [reply]
Mmm... maybe `strace` is not actually useful, as the syscall to open the `libmpi.so` needs to happen only once. Perhaps using the `ltrace` (different command, note the initial `l`), with some appropriate flags I don't know, may shed some extra insight.
